python/ql/src/Security/CWE-327/FluentApiModel.qll

```diff
@@ -0,0 +1,81 @@
+import python
+import TlsLibraryModel
+
+/**
+ * Configuration to track flow from the creation of a context to
+ * that context being used to create a connection.
+ * Flow is broken if the insecure protocol of interest is being restricted.
+ */
+class InsecureContextConfiguration extends DataFlow::Configuration {
+  TlsLibrary library;
+  ProtocolVersion tracked_version;
+
+  InsecureContextConfiguration() {
+    this = library + "Allows" + tracked_version and
+    tracked_version.isInsecure()
+  }
+
+  ProtocolVersion getTrackedVersion() { result = tracked_version }
+
+  override predicate isSource(DataFlow::Node source) {
+    // source = library.unspecific_context_creation()
+    exists(ProtocolUnrestriction pu |
+      pu = library.protocol_unrestriction() and
+      pu.getUnrestriction() = tracked_version
+    |
+      source = pu.getContext()
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink = library.connection_creation().getContext()
+  }
+
+  override predicate isBarrierOut(DataFlow::Node node) {
+    exists(ProtocolRestriction r |
+      r = library.protocol_restriction() and
+      node = r.getContext() and
+      r.getRestriction() = tracked_version
+    )
+  }
+
+  override predicate isBarrierIn(DataFlow::Node node) {
+    exists(ProtocolUnrestriction r |
+      r = library.protocol_unrestriction() and
+      node = r.getContext() and
+      r.getUnrestriction() = tracked_version
+    )
+  }
+}
+
+/**
+ * A connection is created from a context allowing an insecure protocol,
+ * and that protocol has not been restricted appropriately.
+ */
+predicate unsafe_connection_creation(
+  DataFlow::Node creation, ProtocolVersion insecure_version, DataFlow::Node source, boolean specific
```

Comment:
Since we're talking about both _connection_ creation, and _context_ creation, it could be nice to disambiguate.

```suggestion
  DataFlow::Node connectionCreation, ProtocolVersion insecure_version, DataFlow::Node source, boolean specific
```